File: backend/app/services/vaccine_service.py
from datetime import date, datetime, timedelta
from typing import List, Optional, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import and_
from app.models.vaccine_reminder import VaccineReminder
from app.models.pet import Pet


class VaccineService:
    """Serviço para gerenciamento de lembretes de vacinas"""
    
    # Vacinas comuns e seus intervalos padrão (em meses)
    COMMON_VACCINES = {
        "dog": {
            "V8/V10 (Polivalente)": 12,
            "Antirrábica": 12,
            "Gripe Canina": 12,
            "Giárdia": 12,
            "Leishmaniose": 12,
        },
        "cat": {
            "V3/V4/V5 (Polivalente)": 12,
            "Antirrábica": 12,
            "Leucemia Felina (FeLV)": 12,
        }
    }
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_reminder(
        self,
        pet_id: int,
        vaccine_name: str,
        scheduled_date: date,
        owner_id: int,
        notify_days_before: int = 7,
        notes: str = None
    ) -> Tuple[Optional[VaccineReminder], str]:
        """Cria um lembrete de vacina"""
        
        # Verifica se o pet pertence ao usuário
        pet = self.db.query(Pet).filter(
            Pet.id == pet_id,
            Pet.owner_id == owner_id
        ).first()
        
        if not pet:
            return None, "Pet não encontrado"
        
        reminder = VaccineReminder(
            pet_id=pet_id,
            vaccine_name=vaccine_name,
            scheduled_date=scheduled_date,
            notify_days_before=notify_days_before,
            notes=notes
        )
        
        self.db.add(reminder)
        self.db.commit()
        self.db.refresh(reminder)
        
        return reminder, "Lembrete criado com sucesso"
    
    def list_reminders(
        self,
        owner_id: int,
        pet_id: int = None,
        include_completed: bool = False
    ) -> List[VaccineReminder]:
        """Lista lembretes de vacinas"""
        
        query = self.db.query(VaccineReminder).join(Pet).filter(
            Pet.owner_id == owner_id
        )
        
        if pet_id:
            query = query.filter(VaccineReminder.pet_id == pet_id)
        
        if not include_completed:
            query = query.filter(VaccineReminder.is_completed == False)
        
        reminders = query.order_by(VaccineReminder.scheduled_date).all()
        
        # Adiciona campos calculados
        today = date.today()
        for reminder in reminders:
            days_diff = (reminder.scheduled_date - today).days
            reminder.days_until = days_diff
            reminder.is_overdue = days_diff < 0 and not reminder.is_completed
        
        return reminders
    
    def get_upcoming_vaccines(
        self,
        owner_id: int,
        days_ahead: int = 30
    ) -> dict:
        """Retorna vacinas próximas e atrasadas"""
        
        today = date.today()
        future_date = today + timedelta(days=days_ahead)
        
        # Busca todas as vacinas pendentes
        all_reminders = self.db.query(VaccineReminder).join(Pet).filter(
            Pet.owner_id == owner_id,
            VaccineReminder.is_completed == False
        ).order_by(VaccineReminder.scheduled_date).all()
        
        upcoming = []
        overdue = []
        
        for reminder in all_reminders:
            days_diff = (reminder.scheduled_date - today).days
            reminder.days_until = days_diff
            reminder.is_overdue = days_diff < 0
            
            if days_diff < 0:
                overdue.append(reminder)
            elif days_diff <= days_ahead:
                upcoming.append(reminder)
        
        return {
            "upcoming": upcoming,
            "overdue": overdue,
            "total_pending": len(upcoming) + len(overdue)
        }
    
    def complete_reminder(
        self,
        reminder_id: int,
        owner_id: int,
        completed_date: date,
        record_id: int = None
    ) -> Tuple[Optional[VaccineReminder], str]:
        """Marca uma vacina como aplicada"""
        
        reminder = self.db.query(VaccineReminder).join(Pet).filter(
            VaccineReminder.id == reminder_id,
            Pet.owner_id == owner_id
        ).first()
        
        if not reminder:
            return None, "Lembrete não encontrado"
        
        reminder.is_completed = True
        reminder.completed_date = completed_date
        reminder.record_id = record_id
        
        self.db.commit()
        self.db.refresh(reminder)
        
        return reminder, "Vacina marcada como aplicada"
    
    def update_reminder(
        self,
        reminder_id: int,
        owner_id: int,
        **kwargs
    ) -> Tuple[Optional[VaccineReminder], str]:
        """Atualiza um lembrete"""
        
        reminder = self.db.query(VaccineReminder).join(Pet).filter(
            VaccineReminder.id == reminder_id,
            Pet.owner_id == owner_id
        ).first()
        
        if not reminder:
            return None, "Lembrete não encontrado"
        
        for key, value in kwargs.items():
            if value is not None and hasattr(reminder, key):
                setattr(reminder, key, value)
        
        self.db.commit()
        self.db.refresh(reminder)
        
        return reminder, "Lembrete atualizado"
    
    def delete_reminder(
        self,
        reminder_id: int,
        owner_id: int
    ) -> bool:
        """Remove um lembrete"""
        
        reminder = self.db.query(VaccineReminder).join(Pet).filter(
            VaccineReminder.id == reminder_id,
            Pet.owner_id == owner_id
        ).first()
        
        if not reminder:
            return False
        
        self.db.delete(reminder)
        self.db.commit()
        return True
    
    def suggest_vaccines(self, pet_id: int, owner_id: int) -> List[dict]:
        """Sugere vacinas com base na espécie do pet"""
        
        pet = self.db.query(Pet).filter(
            Pet.id == pet_id,
            Pet.owner_id == owner_id
        ).first()
        
        if not pet:
            return []
        
        vaccines = self.COMMON_VACCINES.get(pet.species, {})
        
        # Verifica quais já estão agendadas
        existing = self.db.query(VaccineReminder.vaccine_name).filter(
            VaccineReminder.pet_id == pet_id,
            VaccineReminder.is_completed == False
        ).all()
        existing_names = {r.vaccine_name for r in existing}
        
        suggestions = []
        for name, interval_months in vaccines.items():
            suggestions.append({
                "vaccine_name": name,
                "interval_months": interval_months,
                "already_scheduled": name in existing_names
            })
        
        return suggestions
    
    def get_reminders_to_notify(self) -> List[VaccineReminder]:
        """
        Retorna lembretes que precisam ser notificados.
        Usado por um job/cron para enviar notificações.
        """
        today = date.today()
        
        reminders = self.db.query(VaccineReminder).filter(
            VaccineReminder.is_completed == False,
            VaccineReminder.scheduled_date <= today + timedelta(days=30)
        ).all()
        
        to_notify = []
        for reminder in reminders:
            days_until = (reminder.scheduled_date - today).days
            
            # Notifica se está dentro da janela de notificação
            if days_until <= reminder.notify_days_before:
                # Verifica se já foi notificado hoje
                if reminder.last_notified_at:
                    if reminder.last_notified_at.date() == today:
                        continue
                
                to_notify.append(reminder)
        
        return to_notify
    
    def mark_as_notified(self, reminder_ids: List[int]):
        """Marca lembretes como notificados"""
        self.db.query(VaccineReminder).filter(
            VaccineReminder.id.in_(reminder_ids)
        ).update(
            {"last_notified_at": datetime.utcnow()},
            synchronize_session=False
        )
        self.db.commit()

File: mobile/lib/services/vaccine_service.dart
import 'api_service.dart';
import 'pet_service.dart';
import 'notification_service.dart';

class VaccineReminder {
  final int id;
  final int petId;
  final String vaccineName;
  final DateTime scheduledDate;
  final bool isCompleted;
  final DateTime? completedDate;
  final int? recordId;
  final int notifyDaysBefore;
  final String? notes;
  final int? daysUntil;
  final bool isOverdue;

  VaccineReminder({
    required this.id,
    required this.petId,
    required this.vaccineName,
    required this.scheduledDate,
    required this.isCompleted,
    this.completedDate,
    this.recordId,
    required this.notifyDaysBefore,
    this.notes,
    this.daysUntil,
    this.isOverdue = false,
  });

  factory VaccineReminder.fromJson(Map<String, dynamic> json) {
    return VaccineReminder(
      id: json['id'],
      petId: json['pet_id'],
      vaccineName: json['vaccine_name'],
      scheduledDate: DateTime.parse(json['scheduled_date']),
      isCompleted: json['is_completed'] ?? false,
      completedDate: json['completed_date'] != null 
          ? DateTime.parse(json['completed_date']) 
          : null,
      recordId: json['record_id'],
      notifyDaysBefore: json['notify_days_before'] ?? 7,
      notes: json['notes'],
      daysUntil: json['days_until'],
      isOverdue: json['is_overdue'] ?? false,
    );
  }

  String get statusText {
    if (isCompleted) return 'Aplicada';
    if (isOverdue) return 'Atrasada';
    if (daysUntil != null) {
      if (daysUntil == 0) return 'Hoje';
      if (daysUntil == 1) return 'Amanhã';
      if (daysUntil! <= 7) return 'Em $daysUntil dias';
      return 'Agendada';
    }
    return 'Pendente';
  }
}

class VaccineSuggestion {
  final String vaccineName;
  final int intervalMonths;
  final bool alreadyScheduled;

  VaccineSuggestion({
    required this.vaccineName,
    required this.intervalMonths,
    required this.alreadyScheduled,
  });

  factory VaccineSuggestion.fromJson(Map<String, dynamic> json) {
    return VaccineSuggestion(
      vaccineName: json['vaccine_name'],
      intervalMonths: json['interval_months'],
      alreadyScheduled: json['already_scheduled'] ?? false,
    );
  }
}

class UpcomingVaccines {
  final List<VaccineReminder> upcoming;
  final List<VaccineReminder> overdue;
  final int totalPending;

  UpcomingVaccines({
    required this.upcoming,
    required this.overdue,
    required this.totalPending,
  });

  factory UpcomingVaccines.fromJson(Map<String, dynamic> json) {
    return UpcomingVaccines(
      upcoming: (json['upcoming'] as List)
          .map((r) => VaccineReminder.fromJson(r))
          .toList(),
      overdue: (json['overdue'] as List)
          .map((r) => VaccineReminder.fromJson(r))
          .toList(),
      totalPending: json['total_pending'],
    );
  }

  bool get hasOverdue => overdue.isNotEmpty;
  bool get hasUpcoming => upcoming.isNotEmpty;
}

class VaccineService {
  final ApiService _api = ApiService();

  /// Cria um lembrete de vacina
  Future<VaccineReminder> createReminder({
    required int petId,
    required String vaccineName,
    required DateTime scheduledDate,
    int notifyDaysBefore = 7,
    String? notes,
  }) async {
    final response = await _api.post('/vaccines', {
      'pet_id': petId,
      'vaccine_name': vaccineName,
      'scheduled_date': scheduledDate.toIso8601String().split('T').first,
      'notify_days_before': notifyDaysBefore,
      if (notes != null) 'notes': notes,
    });
    
    final reminder = VaccineReminder.fromJson(response);
    
    // Add Notification
    try {
      final petService = PetService();
      final pet = await petService.getPet(petId);
      final petName = pet.name;
      
      await NotificationService().scheduleVaccineReminder(
        reminderId: reminder.id,
        petName: petName,
        vaccineName: vaccineName,
        scheduledDate: scheduledDate,
        daysBefore: notifyDaysBefore,
      );
    } catch (e) {
      print('Erro ao agendar notificação de vacina: $e');
    }
    
    return reminder;
  }

  /// Lista lembretes de vacinas
  Future<List<VaccineReminder>> listReminders({
    int? petId,
    bool includeCompleted = false,
  }) async {
    var endpoint = '/vaccines';
    final params = <String>[];
    
    if (petId != null) params.add('pet_id=$petId');
    if (includeCompleted) params.add('include_completed=true');
    
    if (params.isNotEmpty) {
      endpoint += '?${params.join('&')}';
    }
    
    final response = await _api.get(endpoint);
    return (response as List)
        .map((r) => VaccineReminder.fromJson(r))
        .toList();
  }

  /// Obtém vacinas próximas e atrasadas
  Future<UpcomingVaccines> getUpcomingVaccines({int daysAhead = 30}) async {
    final response = await _api.get('/vaccines/upcoming?days_ahead=$daysAhead');
    return UpcomingVaccines.fromJson(response);
  }

  /// Obtém sugestões de vacinas para um pet
  Future<List<VaccineSuggestion>> getSuggestions(int petId) async {
    final response = await _api.get('/vaccines/suggestions/$petId');
    return (response['suggestions'] as List)
        .map((s) => VaccineSuggestion.fromJson(s))
        .toList();
  }

  /// Obtém detalhes de um lembrete
  Future<VaccineReminder> getReminder(int reminderId) async {
    final response = await _api.get('/vaccines/$reminderId');
    return VaccineReminder.fromJson(response);
  }

  /// Atualiza um lembrete
  Future<VaccineReminder> updateReminder(
    int reminderId, {
    String? vaccineName,
    DateTime? scheduledDate,
    int? notifyDaysBefore,
    String? notes,
  }) async {
    final data = <String, dynamic>{};
    if (vaccineName != null) data['vaccine_name'] = vaccineName;
    if (scheduledDate != null) {
      data['scheduled_date'] = scheduledDate.toIso8601String().split('T').first;
    }
    if (notifyDaysBefore != null) data['notify_days_before'] = notifyDaysBefore;
    if (notes != null) data['notes'] = notes;
    
    final response = await _api.patch('/vaccines/$reminderId', data);
    final reminder = VaccineReminder.fromJson(response);
    
    // Update Notification
    try {
      if (reminder.scheduledDate.isAfter(DateTime.now())) {
        final petService = PetService();
        final pet = await petService.getPet(reminder.petId);
        final petName = pet.name;
        
        await NotificationService().cancelReminder(reminderId);
        await NotificationService().scheduleVaccineReminder(
          reminderId: reminder.id,
          petName: petName,
          vaccineName: reminder.vaccineName,
          scheduledDate: reminder.scheduledDate,
          daysBefore: reminder.notifyDaysBefore,
        );
      }
    } catch (e) {
      print('Erro ao atualizar notificação de vacina: $e');
    }
    
    return reminder;
  }

  /// Marca vacina como aplicada
  Future<VaccineReminder> completeReminder(
    int reminderId, {
    required DateTime completedDate,
    int? recordId,
  }) async {
    final response = await _api.post('/vaccines/$reminderId/complete', {
      'completed_date': completedDate.toIso8601String().split('T').first,
      if (recordId != null) 'record_id': recordId,
    });
    
    // Cancel any pending notifications
    await NotificationService().cancelReminder(reminderId);
    
    return VaccineReminder.fromJson(response);
  }

  /// Remove um lembrete
  Future<void> deleteReminder(int reminderId) async {
    await _api.delete('/vaccines/$reminderId');
    await NotificationService().cancelReminder(reminderId);
  }
}

